Skolem based gadt constraints

AleksanderBG opened the pull request:
Fixes #5667.
Fixes #2985.

AleksanderBG commented:
@odersky this looks ready to go! Commit messages contain basically all the information. Typeclass derivation needed to be altered, since from:
```scala
(t: T) match {
  case _: U => ...
}
```
we're not allowed to deduce that `U <: T` (consider that if `U = Any`, the pattern always matches).

I realize that approximating Skolem bounds in the way I'm doing it is unsound, but for now it prevents an even worse unsoundness. Exact problem is that `Sko(U) | Sko(U)` is simplified to `U` for two different Skolems, which is not what we want when considering the constraint like `T >: Sko(U) | Sko(U)`. I'm noting for posterity that Paolo suggested we could have a separate mode where we maintain the "necessary" conditions while simplifying unions (and intersections?) - i.e. do not simplify such unions. That approach would solve the problem I've encountered and we might want to implement it later.

smarter commented:
> Exact problem is that Sko(U) | Sko(U) is simplified to U for two different Skolems,

Wait, if the constraint solver does this kind of stuff then surely it's already unsound and we should fix it ?

AleksanderBG commented:
@smarter if I understood @Blaisorblade, that's the difference between "necessary" and "sufficient". What the current constraint solver does is "sufficient", and that is what is needed for type inference. What GADT constraints need is "necessary", and, well, they don't have it (yet). For now, if constraint already contains `T >: S1 <: S1` and we're asked to record that `T >: Sko(U)`, we ensure that `S1 | Sko(U) <:< S2`. This is quite likely unsound, but the only case I can think of where it makes a difference is if we need somehow to check that `T` is inhabited (consider that all we record is a _lower bound_ of a Skolem). I think for now it's fine to just merge this PR, since it gets rid of type unsoundness of this kind:
```scala
final case class Inv[T](t: T)
def foo[T](t: T): T = Inv(t) match { case Inv(_: Int) => 0 }
val notFive: 5 = foo[5](5) // actually zero
```

odersky commented:
I fail to see why simplifying  `T >: Sko(U) | Sko(U)` to `T >: U` is a problem. That's precisely as specified as far as I am concerned. We do check that the resulting constraint is satisfiable. If it is not, the subtyping derivation fails.

odersky commented on compiler/src/dotty/tools/dotc/core/Contexts.scala:
```diff
@@ -779,6 +779,21 @@ object Contexts {
 
+      // avoid recording skolems in lower bounds
+      // recording two skolem bounds results in an union, which is then simplified
+      // T >: Sko(U) | Sko(U) is simplified to T >: U, which is simply wrong
```
Where is this simplification of `T >: Sko(U) | Sko(U)` to `T >: U` done? If this done only for Skolemtypes? I suspect the answer is that it happens for all singleton types. But then the fix should also be generalized to all singleton types, right?

AleksanderBG commented:
test performance please

smarter commented on compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:
```diff
@@ -102,7 +103,7 @@ trait ConstraintHandling[AbstractContext] {
           try
             if (isUpper) oldBounds.derivedTypeBounds(lo, hi & bound)
-            else oldBounds.derivedTypeBounds(lo | bound, hi)
+            else oldBounds.derivedTypeBounds(typeLub(lo, bound), hi)
```
Not sure this is enough in general, there's a few calls to `|` and `&` in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/Constraint.scala (even though the documentation of ConstraintHandling claims "Constraint code does not have access to a type-comparer. Anything regarding lubs and glbs has to be done elsewhere.")

AleksanderBG commented:
Rebased on top of master. More changes to typeclass derivation tests were needed - I'll open a separate PR for replacing erasedValue with typeOf (see #5564, last 3 comments specifically).

@smarter I stared very hard at `OrderingConstraint` and `Constraint`, and I believe that current implementation is sound. Specifically: I think the calls to `TypeComparer#lub` in `fullXBounds` only differ from `OrType` when HK types are involved (see `TypeComparer#liftIfHK`), and the ones in `replace` only rely on `Any | T` being simplified to `T`. I added a test for some more creative usages of GADT bounds in the last commit. I think this PR is fine to merge as-is, though we may want to re-think `Constraint` calling lub/glb later on.

@odersky I believe GADT bounds cannot ever be approximated, only simplified. I _think_ it'd be sound to simplify `T >: Sko(U) | Sko(U)` to `T >: Sko(U)`, but I also think we should wait until we know there's an actual performance problem. Normal compilation times seem to be mostly unaffected.

Blaisorblade commented:
> I _think_ it'd be sound to simplify `T >: Sko(U) | Sko(U)` to `T >: Sko(U)`

If those are the *same* `Sko(U)`, yes, otherwise it's not clear. Let me write `Sko(i, U)` to make the identity `i` explicit.

If `i1 != i2`, `T >: Sko(i1, U)` admits `T = Sko(i1, U)` as solution, while `T >: Sko(i1, U) | Sko(i2, U)` doesn't, so simplifying the latter constraint to the former would add invalid solution. Following @odersky's summary, that's not okay for inference, but okay for GADTs (that is, for proving some constraints are unsatisfiable).

BTW, unsound inference results are usually caught later, but only if the violation is visible in the trees produced by inference. Conversely, `constrainPatternType(SkolemType(subtp), tp)` invents a Skolem and uses it in a constraint, but it's not obvious this constraint is added to to the actual tree.

AleksanderBG commented:
@Blaisorblade I'm less and less clear on whether it's actually sound or not to simplify the bounds in such a manner. For _typechecking_, intuitively it should be acceptable to know _less_ bounds, because then we will simply not accept some "morally valid" code. Note that this'd mean that it'd be sound to drop _both_ skolems from the lower bound without losing soundness. For _exhaustivity_ and useless pattern checks (which you may have meant by "some constraints are unsatisfiable"), we actually would not want to simplify the constraints in such a way (since `T = Sko(i1, U)` might be a constraint required by a subpattern, and knowing that `T >: Sko(i1, U) | Sko(i2, U)` lets us decide that a pattern is useless).

Blaisorblade commented:
> I'm less and less clear on whether it's actually sound or not to simplify the bounds in such a manner. For _typechecking_, intuitively it should be acceptable to know _less_ bounds, because then we will simply not accept some "morally valid" code.
> For _exhaustivity_ and useless pattern checks (which you may have meant by "some constraints are unsatisfiable"), we actually would not want to simplify the constraints in such a way (since `T = Sko(i1, U)` might be a constraint required by a subpattern, and knowing that `T >: Sko(i1, U) | Sko(i2, U)` lets us decide that a pattern is useless).

Not approximating is better than approximating if we can afford it, of course. So if you can avoid simplifying those skolems without performance loss, IMHO go ahead.

But you talk about soundness, and I disagree there. And I think the two scenarios you describe should work the same way.

Overall, we seem to have three scenarios:
1. type inference, where the solver must search for *sufficient* constraints
2. GADT typechecking, where the solver must search for *necessary* constraints; this isn't implemented yet, but let me assume it'll be.
3. warnings about missing branches (exhaustivity) or useless branches (redundant patterns.

It seems the solver can only be in sufficient or (if implemented) necessary mode.
For scenario 3, we can pick which — since we must always approximate in one or the other direction. Neither direction breaks type soundness, but missing branches cause `MatchError` while extra branches are just extra code (which can be `assert(false)` if the user is sure).
But the warning says "match may not be exhaustive" — that is, the compiler couldn't prove the missing cases are impossible, so it approximate to "please add them".
That is, a pattern is requested *unless we can soundly prove it's impossible*. That's what I argued in #4029.

If we agree on this behavior for scenario 3, then the solver must search for *necessary* constraints, like scenario 2 — and if the resulting constraint is unsatisfiable, the original ones were as well and the considered branch is impossible.

The above implies the compiler might require unused cases. You seem to disagree.
But we can't avoid errors in both directions, so do you want to err in the other direction?

We should pick somehow and stick to that.

AleksanderBG commented:
@Blaisorblade that comment was from two weeks ago, before we discussed everything thoroughly and I came to the conclusions from my presentation/report. The situation, as I understand it now, is as follows:

_Type inference_ is concerned with finding a solution satisfying constraint `C`. We can instead solve a constraint `C'` (found, for example, by simplifying `T >: U(x) | U(y)`* to `T >: U`) iff a solution to `C'` is guaranteed to be a solution for `C`. This is what I understand to be the _sufficient_ case - solving `C'` is sufficient for solving `C`.

\* \- allow me to use the syntax `U(x)` for denoting some singleton of type `U`, since singleton types other than skolems concern us as well.

_GADT constraint inference_ is concerned with finding the strongest possible constraint `D` entailed by `C, G, x: P, x.type <: S` where `C` is the current constraint, `G` is the current environment, `x` is an existential term of types both `P` and `S`. If `D` is entailed by the above, then we can return `D'` instead as long as any solution to `D` is also a solution to `D'`. This is the _necessary_ case - solving `D'` is necessary for solving `D`. We can also say that `D'` is no stronger that `D`, which in a roundabout way brings us to the "less constraints is always sound for typechecking" argument.

To _emit safe GADT exhaustivity warnings_, we must decide if the constraint `C` entailed by a particular missing pattern is unsatisfiable. Yes, we want/should err on the side of caution and emit a warning even if we are uncertain that the pattern is necessary. Therefore, instead of solving `UNSAT(C)`, we may solve `UNSAT(C')` iff any solution to `C` is guaranteed to be a solution to `C'`. Again, this is the _necessary_ case.

What this means is that for GADT constraints, we can simplify `T >: U(x) | U(y)` to `T >: U(x)` _or_ `T >: U(y)` _or_ `T >: Nothing`. However! For constraint inference, this would mean that we don't accept code that is morally valid. For exhaustivity checking, this would mean that we emit spurious warnings. Both of those cases are bad from user perspective - we should have a good reason to not accept code that should compile or to bother users with useless warnings. Since there's so far no observable impact on performance, I'd just not simplify the constraint for the time being. There's some subtlety to that argument, but I'd very much like to first finish fixing the errors in this PR that I discovered and then to present everything wrapped up together.

AleksanderBG commented on compiler/src/dotty/tools/dotc/core/TypeComparer.scala:
```diff
@@ -403,6 +408,16 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] {
+                  // one of which is constrained to be a subtype of another.
+                  // We do not need similar code in fourthTry, since we only need to care about
+                  // comparing two constrained types, and that case will be handled here first.
+                  ctx.gadt.isLess(tp1.symbol, tp2.symbol) && GADTusage(tp1.symbol) && GADTusage(tp2.symbol)
```
@smarter you might want to take a look at this and its original commit (the first in history). It seems to me that we might be missing an analogous case when comparing two TypeParamRefs in frozen constraint. Note that when _not_ in frozen constraint, `TypeComparer` will try to record a new bound, which will be a no-op.

AleksanderBG replied:
Reminder to self post-coversation w/ @smarter: create an issue for this after this is merged.

AleksanderBG commented:
Status update: I am aware of one remaining problem in `TypeComparer` - what constraints to infer when `A,B` constrainable and `T <:< A | B`? I believe the answer to be: constraint `C` s.t. `T <: A` entails `C` _and_ `T <: B` entails `C`. This requires some sort of constraint _intersection_ (Paolo called this an "or"). Currently we infer that `T <: A, T <: B` which easily leads to unsoundness:
```scala
enum SUB[A, +B] { case Refl[T]() extends SUB[T, T] }
def foo[A, B](sub: Int SUB (A|B)): A =
  sub match { case SUB.Refl() => 0 }
foo[String, Int](SUB.Refl()) // ClassCastException
```

smarter commented:
Just FYI since I happened to stumble upon it, here's another place where constraints might be cut:
https://github.com/lampepfl/dotty/blob/df8f79f6b9bb65aa2ddd7fc9433df4ba669fb8d1/compiler/src/dotty/tools/dotc/core/TypeComparer.scala#L1561-L1562

odersky commented on compiler/src/dotty/tools/dotc/typer/Implicits.scala:
```diff
@@ -397,21 +397,29 @@ object Implicits {
-            mapOver(t)
+      val ctx0 = ctx
+      locally {
+        implicit val ctx = ctx0.fresh.setTyperState {
```
It seems you can simply write
```
   implicit val ctx = ctx0.fresh.setExploreTyperState()
```
for that. The constraint is copied anyway. 

Also, it's more idiomatic to pass the new constraint to a method. I.e. instead of ctx0/locally, you could write:

```
  def map(implicit ctx: Context) = ...
  map(ctx0.fresh.setExploreTyperState())
```

AleksanderBG replied:
Simplified as suggested.

odersky commented on compiler/src/dotty/tools/dotc/typer/Inferencing.scala:
```diff
@@ -153,41 +153,22 @@ object Inferencing {
    *
-   *       StaticScrutineeType           PatternType
-   *                         \            /
-   *                      DynamicScrutineeType
```
I think this diagram is useful info to have. Is it wrong? If not let's keep it.

AleksanderBG replied:
Restored the old comment.

odersky commented on compiler/src/dotty/tools/dotc/typer/Inferencing.scala:
```diff
@@ -153,41 +153,22 @@ object Inferencing {
+   * be matching on an expression as well, we take a skolem of the scrutinee, which is essentially an existential
+   * singleton type (see [[dotty.tools.dotc.core.Types.SkolemType]]).
    *
-   *            StaticScrutineeType
```
Same here. This is tricky stuff. Let's keep all info unless it is wrong.

odersky commented on compiler/src/dotty/tools/dotc/typer/Namer.scala:
```diff
@@ -1214,8 +1214,16 @@ class Namer { typer: Typer =>
-      var rhsCtx = ctx.addMode(Mode.InferringReturnType)
+      var rhsCtx = ctx.fresh.addMode(Mode.InferringReturnType)
       if (sym.isInlineMethod) rhsCtx = rhsCtx.addMode(Mode.InlineableBody)
+      if (typeParams.nonEmpty) {
```
Let's refactor this into a common method that is also called from `typedDefDef`.

odersky replied:
I did not understand why we need to add type parameters to GADT bounds here (or in typedDefDef). This should be documented.

AleksanderBG replied:
Comments were added to both places.

odersky commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -1535,19 +1506,38 @@ class Typer extends Namer
-        rhsCtx.gadt.addBound(tdef.symbol, tr, isUpper = true)
+    val rhsCtx: Context = {
+      var _result: FreshContext = null
+      def resultCtx(): FreshContext = {
```
Can we factor this better? I am getting a bit frustrated how much complexity is increased by the GADT stuff. Right now it is a tradeoff: Slightly more expressive GADTs is bought by a significant drop in maintainability of the compiler as a whole (yes it is that bad!). We should not be forced to make that tradeoff.

AleksanderBG replied:
This block was more complex than the one before, because I tried to not unnecessarily instantiate  a fresh context. I reverted it to the old version and further simplified both branches to just a single call to a GadtConstraint method.

AleksanderBG commented on compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:
```diff
@@ -27,10 +30,13 @@ trait ConstraintHandling[AbstractContext] {
   protected def constraint: Constraint
   protected def constraint_=(c: Constraint): Unit
 
+  protected def externalize(param: TypeParamRef)(implicit ctx: Context): Type
```
Removed after simplifications - we now just override `xBounds` in `GadtConstraint`.

AleksanderBG commented on compiler/src/dotty/tools/dotc/core/Constraint.scala:
```diff
@@ -45,6 +45,12 @@ abstract class Constraint extends Showable {
   /** The parameters that are known to be greater wrt <: than `param` */
   def upper(param: TypeParamRef): List[TypeParamRef]
 
+  /** `lower`, except that `minLower.forall(tpr => !minLower.exists(_ <:< tpr))` */
```
Explained.

AleksanderBG commented:
I added recaps of what we talked about in person to all the discussions. Compared to when we talked, the only non-trivial change not related to any discussion is adding `GadtConstraint.addToConstraint(_: List[Symbol])`, which mirrors `ConstraintHandling#addToConstraint` and correctly allows registering symbols with interdependencies. This removed some code duplication between all the places that needed to register multiple symbols in a loop.

odersky commented on compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala:
```diff
@@ -208,7 +208,10 @@ class PlainPrinter(_ctx: Context) extends Printer {
           val bounds =
-            if (constr.contains(tp)) constr.fullBounds(tp.origin)(ctx.addMode(Mode.Printing))
+            if (constr.contains(tp)) {
+              val ctx0 = ctx.addMode(Mode.Printing)
```
Why not leave it inline? `ctx.addMode(Mode.Printing).typeComparer.fullBounds(tp.origin)`
